### PR TITLE
[feature] Added multitenant capablities to django-filters #346

### LINKED ISSUE #346: [bug] Multi-tenant capabilities missing for `django-filters`
I think the `openwisp-users` have [implemented multi-tenant capabilities for DRF views, serializers, and Django admin filters](https://github.com/openwisp/openwisp-users#django-rest-framework-mixins), but they do not have it for [django-filters](https://django-filter.readthedocs.io/en/stable/).

This bug is spotted here: https://github.com/openwisp/openwisp-controller/pull/748#issuecomment-1476062204

### PR DESCRIPTION
**Pull request by Aryamanz29**
**Todo**

- [x] Add tests.
- [x] Add docs.

Closes #346

### CODE REVIEW

File: README.rst [hunk 1]
```diff
@@ -977,6 +977,43 @@ which is inheriting any of the mixin classes.
 Usage example: `organization_field
 <https://github.com/openwisp/openwisp-users#organization_field>`_.
 
+Django filters multi-tenant capabilities for the browsable web UI
```
Comment (nemesisdesign) [inline README.rst:980]:
```suggestion
Multi-tenant filters for the browsable web UI
```

File: README.rst [hunk 2]
```diff
@@ -977,6 +977,43 @@ which is inheriting any of the mixin classes.
...
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Integration of `Django filters <https://django-filter.readthedocs.io/en/stable/guide/rest_framework.html>`_
+with `Django REST Framework <https://www.django-rest-framework.org/>`_
+is provided through a DRF-specific ``FilterSet`` and a ``filter backend``.
+
+The relationship fields in ``django-filters`` options interface shows all the
```
Comment (nemesisdesign) [inline README.rst:987]:
```suggestion
The relationship fields of ``django-filters`` show all the available results,
```

File: README.rst [hunk 3]
```diff
@@ -977,6 +977,43 @@ which is inheriting any of the mixin classes.
...
+relationships, without filtering by the organization the user has access to, which
```
Comment (nemesisdesign) [inline README.rst:988]:
```suggestion
 without filtering by the organization the user has access to, which
```

File: openwisp_users/api/mixins.py [hunk 4]
```diff
@@ -212,6 +214,90 @@ class FilterSerializerByOrgOwned(FilterSerializerByOrganization):
...
+            )
+        if (
+            isinstance(field, ManyToManyField) and field.name != 'user'
+        ):  # pragma: no cover
+            return DjangoOrganizationM2MFilter(
+                queryset=field.remote_field.model.objects.all(),
+                label=field.verbose_name.capitalize(),
+                field_name=field.name,
+                user_attr=cls._user_attr,
+            )
```
Comment (nemesisdesign) [inline openwisp_users/api/mixins.py:273]: there's quite a bit of repetition here, let's improve this. Eg:

```python
opts = dict(
      queryset=field.remote_field.model.objects.all(),
      label=field.verbose_name.capitalize(),
      field_name=field.name,
      user_attr=cls._user_attr
)
not_user = field.name != 'user'

if isinstance(field, ForeignKey) and not_user:
    return DjangoOrganizationFilter(**opts)
if isinstance(field, ManyToManyField) and not_user:
    return DjangoOrganizationM2MFilter(**opts)
```

File: openwisp_users/api/mixins.py [hunk 5]
```diff
@@ -212,6 +214,90 @@ class FilterSerializerByOrgOwned(FilterSerializerByOrganization):
...
+        if isinstance(field, ForeignKey) and field.name != 'user':
+            return DjangoOrganizationFilter(
+                queryset=field.remote_field.model.objects.all(),
+                label=field.verbose_name.capitalize(),
+                field_name=field.name,
+                user_attr=cls._user_attr,
+            )
+        if (
+            isinstance(field, ManyToManyField) and field.name != 'user'
+        ):  # pragma: no cover
```
Comment (nemesisdesign) [inline openwisp_users/api/mixins.py:267]: Red flag :red_square: 

We have a test project where you can add an m2m relationship and write a test to cover this: https://github.com/openwisp/openwisp-users/blob/master/tests/testapp/models.py

File: README.rst [hunk 6]
```diff
@@ -977,6 +977,43 @@ which is inheriting any of the mixin classes.
...
+Usage example:
+
+.. code-block:: python
+
+   from openwisp_users.api.mixins import FilterDjangoByOrgManaged
+   from ..models import FloorPlan
+
+
+   class FloorPlanOrganizationFilter(FilterDjangoByOrgManaged):
+       organization_slug = filters.CharFilter(field_name='organization__slug')
```
Comment (nemesisdesign) [inline README.rst:1003]: in the example here, the import of `filters` is missing.

File: openwisp_users/api/mixins.py [hunk 7]
```diff
@@ -212,6 +214,90 @@ class FilterSerializerByOrgOwned(FilterSerializerByOrganization):
...
+        organization_filter = getattr(user, self._user_attr)
+        # if field_name organization then just organization_filter
+        if self.field_name == 'organization':
+            return queryset.filter(pk__in=organization_filter)
+        # for field_name other than organization
+        conditions = Q(**{'organization__in': organization_filter})
+        return queryset.filter(conditions)
+
+    def __init__(self, *args, **kwargs):
+        self._user_attr = kwargs.get('user_attr')
```
Comment (nemesisdesign) [inline openwisp_users/api/mixins.py:235]:
```suggestion
        self._user_attr = kwargs.pop('user_attr')
```

And remove the next line.

**Review by nemesisdesign — CHANGES_REQUESTED** [review]
Great progress @Aryamanz29! :clap: 

I did some testing and it's working well.
There's some room for improvement, since this is going to become a feature we will use in all our modules.

I believe we should add something like this to this library to avoid repeating the same over and over in all modules:

```python
class BaseOrganizationFilter(FilterDjangoByOrgManaged):
   organization_slug = filters.CharFilter(field_name='organization__slug')
   class Meta:
       fields = ['organization', 'organization_slug']
```

Usage example in Network Topology module:
```
class OrganizationFilter(BaseOrganizationFilter):
    class Meta(BaseOrganizationFilter.Meta):
        model = Node
```

See also my comments below.

File: openwisp_users/api/mixins.py (hunk 5 above)
Comment (Aryamanz29) [reply openwisp_users/api/mixins.py]: While generating test coverage locally, I accidentally left this comment in the code and forgot to remove it. My bad, and thank you for bringing it to my attention 👍

**Review by nemesisdesign — APPROVED** [review]

**Pull request merged**